src/v/ssx/work_queue.cc

```diff
@@ -0,0 +1,37 @@
+// Copyright 2023 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#include "ssx/work_queue.h"
+
+#include <exception>
+
+namespace ssx {
+work_queue::work_queue(error_reporter_fn fn)
+  : _error_reporter(std::move(fn)) {}
+
+void work_queue::submit(ss::noncopyable_function<ss::future<>()> fn) {
+    if (_as.abort_requested()) {
+        return;
+    }
+    _tail = _tail
+              .then([this, fn = std::move(fn)]() {
+                  if (_as.abort_requested()) {
+                      return ss::now();
+                  }
+                  return fn();
```

Comment:
Possibly, I'm not really sure how you would explain `the resulting futures are waited on in order as a chain` semantically, what would that mean?

Reply:
it's common to build async functions with seastar like this

```c++
future<> foo() {
  <some code executed uninterrupted with the caller>
  co_await ...
}
```

where the first part of `foo()` is some kind of bookkeeping that shouldn't be interleaved, but makes logical sense to have in `foo()` and not the caller. seastar ensure this behavior.

so with respect to my comment, it would have been something like

```c++
auto f = foo(); // f is future<> not void
queue.insert(std::move(f));
```

but now I'm realizing I'm suggesting something entirely different than

```c++
queue.insert([] () -> future<> {});
```

so i think you're right just totally different things I was thinking about.

Reply:
I feel like maybe you could maybe still do that by doing something like:

```
ss::future<> f = foo();
queue.submit([f = std::move(f)]() mutable { return std::move(f); });
```

But I still feel like that is tricky to reason about, as I'm not sure you get any ordering guarantees.. Maybe if you have a concrete example I could noodle on it.

Reply:
yep that would also work. no concrete example comes to mind